packages/smithy-client/src/parse-utils.ts

```diff
@@ -262,19 +255,28 @@ export const strictParseFloat = strictParseDouble;
  * @returns The value as a number, or undefined if it's null/undefined.
  */
 export const strictParseFloat32 = (value: string | number): number | undefined => {
-  if (value === "NaN") {
-    return NaN;
-  }
   if (typeof value == "string") {
-    const parsed: number = parseFloat(value);
-    if (Number.isNaN(parsed)) {
-      throw new TypeError(`Expected real number, got implicit NaN`);
-    }
-    return expectFloat32(parsed);
+    return expectFloat32(parseNumber(value));
   }
   return expectFloat32(value);
 };
 
+// This regex matches JSON-style numbers. In short:
+// * The integral may start with a negative sign, but not a positive one
+// * No leading 0 on the integral unless it's immediately followed by a '.'
+// * Exponent indicated by a case-insensitive 'E' optionally followed by a
+//   positive/negative sign and some number of digits.
+// It also matches both positive and negative infinity as well and explicit NaN.
+const NUMBER_REGEX = /(-?((0(\.\d+)?)|(([1-9]\d*(\.\d)?)\d*)((e|E)(\+|-)?\d+)?))|(-?Infinity)|(NaN)/g;
+
+const parseNumber = (value: string): number => {
+  const matches = value.match(NUMBER_REGEX);
+  if (matches === null || matches[0].length !== value.length) {
```

Comment:
Can you anchor the regex so you can just use `RegExp.test` and not have to check the length of the match?

Reply:
hahaha your faith in js is heartwarming, no it'll match partials anyway